README.md
<p align="center">
  <img src="images/logo.png">
</p>

## About The Project
EEGAILab is a novel machine learning framework designed to analyze and visualize dipole sources from resting electroencephalography data. This analysis uses spectrum analysis and machine learning to capture key frequencies and dipole sources that contribute to classifying specified groups.

For example, we apply this framework to predict long-term treatment outcomes of spinal cord stimulation for patients with chronic lower back pain. We differentiate and show two groups of patients that have undergone spinal cord stimulation. Listed under the expected outputs we show example outputs of our framework including dipole maps and power (uV^2/Hz) differences. In this example, this framework will be able to provide clinicians with a non-invasive technique to predict which patients respond the best to spinal cord stimulation. Below is a figure detailing the multi-stage framework.

<p align="center">
  <img src="images/framework.png">
</p>

Each stage (1-5) has an intermediate output in the `processed_data` folder. The details and input/output of each stage is listed below.
- **<u>EEG Data</u>:** This folder expects preprocessed EEG data. This data requires independent component analysis (ICA) including selection of components and to have been fitted with dipoles. Our example application has been bandpassed from 1-100Hz and truncated into 6 second epochs.

- **<u>Stage 1</u>:** EEG data is first converted to the power spectral density (PSD) with units uV^2/Hz using a fourier transform within EEGLAB. Each whole number frequency (i.e. 1Hz, 2Hz, 3Hz, ...) will be used as a feature in the machine learning algorithm. Data is then saved as a .mat file to speed up computations and accessibility.

- **<u>Stage 2</u>:** The PSD is organized into an excel sheet where rows are subjects and columns are single frequencies to make it easily readable for machine learning. Each data point is the power in uV^2 at that frequency.

- **<u>Stage 3</u>:** SVM models are generated using the frequencies as features. A total of 6 models are made according to the powerbands: `All: 1-100Hz`, `Delta: 1-4Hz`, `Theta: 4-8Hz`, `Alpha: 8-13Hz`, `Beta: 13-30Hz`, and `Gamma: 30-100Hz`. For example, the beta powerband model will only have frequencies 13 to 30Hz. SVM models are trained with a default of 10 iterations of 3-fold cross-validation. This can be changed in `main.m`.

- **<u>Stage 4</u>:** A detailed spreadsheet is compiled about each component's XYZ coordinates in MNI space, location in brain region (hemisphere, lobe, and gyrus), and residual variance. A ranking system was used to rank components based on feature weight and related power. The power, score, and rank of each component is also included.

- **<u>Stage 5</u>:** Power and dipole visuals are generated based on the top 10 ranked components from each subject. The components are the most important dipoles based on the SVM models. A power comparison between the groups is calculated onto an excel spreadsheet while the same dipoles are plotted on a standard MRI brain.

### Subdirectories
- `/data` Folder for pre-processed EEG data. Needs to be added.
- `/images` Images used in README.
- `/processed_data` Intermediate results of the pipeline organized by stages.
- `/src` Source code including functions, helper files, and toolboxes.

## Getting Started
### Prerequisites
The following is a list of items and links with how to install them. The framework will work the best with the specific version numbers. Plugins are installed with EEGLAB.
* [MATLAB 2019b](https://www.mathworks.com/products/matlab.html)
* [LIBSVM v3.24](https://github.com/cjlin1/libsvm/tree/v324)
* [EEGLAB v2021.1](https://github.com/sccn/eeglab/tree/eeglab2021.1)
* ICLabel v1.3.*
* dipfit v4.3.*
* SASICA v1.3.8.*

**EEGLAB Plugin*

### LIBSVM Installation
> Lab members can skip to **step 4** as the toolboxes are included in the Dropbox.
1. Navigate to the LIBSVM GitHub (v324) [link](https://github.com/cjlin1/libsvm/tree/v324).
2. **Download the ZIP** from the **Code** dropdown menu.
3. Create a folder named **toolbox** in the source folder and unzip LIBSVM.
```matlab
src/toolbox/libsvm-3.24
```
4. Add the LIBSVM matlab folder path by right-clicking the folder in MATLAB's directory and selecting `Add to Path > Selected Folders`

<p align="center">
  <img src="images/addpath.png", width="700", height=auto>
</p>

* For Windows users, the required pre-built 64-bit mex file for `'svmtrain.mex'` is in `libsvm-3.24/windows/`
1. Copy `'svmtrain.mex'` to the matlab directory.
```matlab
libsvm-3.24/matlab/
```
* For Mac users, open MATLAB and navigate to the following LIBSVM folder as the current directory.
```matlab
 libsvm-3.24/matlab/
```
1. Type and run `>> make` in the Command Prompt. If it does not work, try `>> mex -setup` to choose a compiler for mex, then type `>> make` for installation. If problems persist refer to the README in the same folder.
2. Mex files `'libsvmread.mex'`, `'libsvmwrite.mex'`, `'svmtrain.mex'`, and `'svmpredict.mex'` should be built after a successful installation.

### EEGLAB (+Plugins) Installation
> Lab members can skip to **step 3** as the toolboxes are included in the Dropbox.
1. Navigate to the EEGLAB toolbox from the GitHub (eeglab2021.1) [link](https://github.com/sccn/eeglab/tree/eeglab2021.1).
2. **Download the ZIP** from the **Code** dropdown menu. (No plugins are included with this approach)
3. Open MATLAB and navigate to the EEGLAB folder as the current directory.
4. Type and run `>> eeglab` in the Command Prompt to add necessary paths and open EEGLAB.
5. To install plugins go to `File > Manage EEGLAB extensions`
6. Select a plugin and click the **Install/Update** button.
    - ICLabel v1.3.
    - dipfit v4.3.
    - SASICA v1.3.8.
<p align="center">
  <img src="images/plugins.png">
</p>

## Usage
Expected runtime: 5 minutes
1. Start MATLAB.
2. Navigate to the EEGAILab folder.
3. Create a `/data` folder and add preprocessed EEG data.
3. Open and run `main.m`.

> Lab members can skip **step 3** as there is an existing data set for validation.

The code can be run as is with preprocessed EEG data or can be adjusted for different testing settings. There are several adjustable variables in the main file which are listed below.

* **Subject IDs**: A vector of unique IDs for each subject. Ex. `[1,2,3,4,5]`
* **Binary Class Labels**: A vector of associated class labels as 1 and -1. Ex. `[1,-1,1,-1,1]`
* **Overwrite**: Overwrite data in stages 1-3 with 'Y' and keep data with 'N' when rerunning code. Ex. `'Y'`
* **Top Number of Components**: The number of components used per subject in a power comparison. Ex. `3`
* **SVM Parameters**
  * **N**: The number of iterations for cross-validation. *Default = `10`*
  * **n_fold**: The number of folds in cross-validation repeated *N* times. *Default = `3`*
  * **feature_sel**: Use feature selection `['N'/'Y']`. *Default = `'N'`*
  * **pval**: P-value for when feature selection is used. *Default = `1`*
  * **weight**: Type of weight used in ranking `['final'/'cv']`. *Default = `'final'`*

### Expected Outputs
Below is an example output when comparing spinal cord stimulation responders using several views (axial, coronal, sagittal) along with the power in uV^2/Hz. The dipole visualization represents the top 10 ranked dipoles from each subject as transparent circles separated by frontal (red) and parietal (blue) lobe. The larger and more opaque a circle the higher contribution it had in the machine learning algorithm. The dipole power shows the average power from the top 3 ranked dipoles from each subject within each group.

<p align="center">
  <img src="images/expectedresults.png", width="85%">
</p>

## Roadmap
- [x] Add Changelog
- [ ] Add Component Selection Guide via EEGLAB

## Contact
For any questions, email **kylebsee@ufl.edu**

### If you use this code or part of this code for your own work, please cite the following publications:

Kyle B. See, Rachel Ho, Stephen Coombes, Ruogu Fang, “TL1 Team Approach to Predicting Response to Spinal Cord Stimulation for Chronic Low Back Pain”, Journal of Clinical and Translational Science, Mar 2021.

Kyle B. See, Rachel Louise Mahealani Judy, Stephen Coombes, Ruogu Fang, “TL1 Team Approach to Predicting Short-term and Long-term Effects of Spinal Cord Stimulation” Journal of Clinical and Translational Science, Jul 2020.

### Useful resources
* [ICLabel Component Labeling Practice](https://labeling.ucsd.edu/tutorial)

## Last Updated
January 7th, 2022

changelog.md
## [0.0.2] - 2022-06-20
### Changed
- `train_svm.m` now uses separate training and testing variables, x_train and x_test, instead of referencing with training and testing indices, train_ind and test_ind.

## [0.0.1] - 2022-01-06
### Added
- `changelog.md`
- `main.m` now creates the *"processed_data"* folder in addition to the *"stage"* subfolders.
- `README.md` now accounts for Windows and Mac differences.
- `README.md` now lists plugins in the EEGLAB (+Plugin) Installation section.
- `README.md` now lists subdirectories and their functions.
- `README.md` now has expected outputs.
- `README.md` now has contact details and references.
### Changed
- `/src` has improved directory reading which now only reads necessary files.
- `main.m` now adds relevant paths needed from `/src`.
- `README.md` descriptions have been updated.
### Fixed
- `strcat` replaced with `fullfile` in `src` for Mac-compatibility when creating and referencing directories.

## [0.0.0] - 2022-01-03
### Added
- EEGAILab code and tools.

main.m
clc;clear;
%% Description
% The current use of the code is to classify between high- and
% moderate-responders of spinal cord stimulation using linear SVM. EEGLAB
% is also used to provide power and dipole visualizations to distinguish 
% differences between responders. All results including intermediate data
% are saved in the same folder under processed_data. For different use,
% modify the variables below.
% 
%   Kyle See 7/1/2021
%   Smart Medical Informatics Learning and Evaluation (SMILE)
%   Biomedical Engineering
% 
% INSTRUCTIONS
% - Place data in the data folder and run this script.
% - Adjust variables if needed
tic
%% Adjustable Variables
% SUBJECT IDS
% - Unique ID for each subject
% - Ex. [1,2,3,4,5,6,7]
subj = [1,2,4,5,7,8,9];

% BINARY CLASS LABELS
% - Vector must be same size as subject ID
% - Use 1 and -1 for labels
% - Ex. [1,-1,-1,1,1,-1,1]
labels = [1,1,-1,-1,1,-1,-1];

% OVERWRITE
% ['Y'/'N'] Default is 'Y'. 'Y' overwrites data in stages 1-3.
overwrite = 'Y';

% TOP NUMBER OF COMPONENTS
% Set the number of top ranking components to consider in a power
% comparison. Default is 3.
top = 3;

% SVM PARAMETERS
N = 10; % Number of iterations to repeat cross-validation. Default is 10.
n_fold = 3; % Number of folds for cross-validation. Default is 3.
feature_sel = 'N'; % Use feature selection ['N'/'Y'].
pval = 1; % P-value for feature selection.
weight = 'final'; % Type of weights to use for ranking ['final'/'cv'].

%% Initialize directory and call functions
% Make folders for stages 1-5 in the processed data folder.
if(~isdeployed)
  cd(fileparts(which('main.m')));
end

% Directory for processed data
path_inter = 'processed_data';
mkdir(path_inter);

for i = 1:5
    folder = fullfile(path_inter, ['stage',num2str(i)]);
    if ~exist(folder,'dir')
        mkdir(folder)
    end
end

% Set paths
path_eeg = 'data'; % EEGLAB data (.set and .fdt)
path_mat = fullfile(path_inter,'stage1'); % MATLAB data (.mat)
path_freq = fullfile(path_inter,'stage2'); % PSD Frequencies (.xlsx)
path_ML = fullfile(path_inter,'stage3'); % Machine learning models
path_comp = fullfile(path_inter,'stage4'); % Component sheets (.xlsx)
path_vis = fullfile(path_inter,'stage5'); % Component visualization (.svg and .xlsx)

% ---------------------------------------------------------------------
% Calling functions
addpath('src/functions')
addpath('src/references')

% 1. Convert EEGLAB data to power spectrum density as .mat file
eeglab2mat(path_eeg,path_mat,overwrite);

% 2. Saving .mat file as machine learning readable excel file
mat2excel(path_mat,path_freq,subj,'freq_features.xlsx',overwrite);

% 3. Perform SVM machine learning
svm_eeg(path_freq,path_ML,labels,N,n_fold,feature_sel,pval,overwrite);

% 4. Get component information and save as an excel file
eeglab2comp(path_eeg,path_ML,path_comp,subj,weight);

% 5a. Visualize and obtain power of components
comp2pwr(path_comp,path_vis,labels,subj,top)

% 5b. Visualize components across axial, coronal, and sagittal views
comp2vis(path_comp,path_vis,labels,subj)
toc

src/functions/comp2pwr.m
function [] = comp2pwr(input_path,output_path,labels,subj,N)
%% Extracts power from IC components based on rank (Stage4 -> Stage5)
% PURPOSE: Take the N number of top ranking components and averages
% component power across subjects within-group. Power is separated between
% high and moderate responders.
%
%   Kyle See 7/1/2021
%   Smart Medical Informatics Learning and Evaluation (SMILE)
%   Biomedical Engineering, University of Florida
%
%   Input:
%       input_path       - Path containing excel component sheets.
%       output_path      - Path to save excel power information.
%       labels           - 1D vector of labels to separate high and
%                          moderate responders. High responders are 1 and
%                          moderate responders are -1.
%
%       Optional Arguments
%       subj             - 1D vector of IDs for each subject. Defaults to 1
%                          to N IDs (1, 2, 3, 4, 5, ..., N).
%       N                - Number of top ranking components to consider
%                          from each subject. Default is 3.
%   Output:
%                        - Generates .xlsx file containing average power
%                          from the top N ranked components. Average power
%                          is found from each powerband, between high and
%                          moderate responders, and between frontal and
%                          parietal lobes.
%
%---------------------------------------------
% Last Updated: 7/1/21
% - 7/1 Created

%% Initializing variables and checking inputs
high_responders = subj(labels == 1); % Separating subjects into groups
moderate_responders = subj(labels == -1); % Separating subjects into groups
dir_xlsx = dir(fullfile(input_path,'*.xlsx')); % Grab directory for .xlsx component sheets (Stage 4)
power_info = []; % Initialize power information

% Check subject numbering and if correct size
if ~exist('subj','var')
    subj = (1:N)';
else
    % Transform to column vector
    if ~isrow(subj)
        subj = subj';
    end
end

% Number of components
if ~exist('N','var')
    total_count = 3;
else
    total_count = N;
end

%% Code
% Loop through each powerband sheet
for ii = 1:length(dir_xlsx)
    
    % Grab name of file ("BAND_components.xlsx")
    filename = dir_xlsx(ii).name;
    name = extractBefore(filename,'_');
    
    % Entire contents of the component sheet here
    data = readtable(fullfile(input_path,filename));
    
    % Separating frontal and parietal data
    data_frontal = data(strcmp(data.lobe,'Frontal'),:);
    data_parietal = data(strcmp(data.lobe,'Parietal'),:);
    
    
    subject_list = unique(data.subject);
    
    % Finding top 3 ranked dipoles
    for i = 1:length(subject_list)
        
        % Find how many frontal components
        comp_num_frontal = length(find(data_frontal.subject == subject_list(i) ));
        
        % Reduces dipoles based on rank if total number per subject exceeds 3
        if comp_num_frontal > total_count
            for j = 1:comp_num_frontal - total_count
                [M,I] = max(data_frontal.(strcat(name,'_rank'))(data_frontal.subject == subject_list(i)));
                data_frontal(I+((i-1)*total_count),:) = [];
            end
        end
        
        % Find how many parietal components
        comp_num_parietal = length(find(data_parietal.subject == subject_list(i) ));
        
        % Reduces dipoles based on rank if total number per subject exceeds 3
        if comp_num_parietal > total_count
            for j = 1:comp_num_parietal - total_count
                [M,I] = max(data_parietal.(strcat(name,'_rank'))(data_parietal.subject == subject_list(i)));
                data_parietal(I+((i-1)*total_count),:) = [];
            end
        end
    end
    
    % Average powers between high and moderate responders separated by
    % frontal and parietal.
    high_res_frontal = [];
    high_res_parietal = [];
    mod_res_frontal = [];
    mod_res_parietal = [];
    
    % Collect all powers from every top ranked component
    for k = 1:length(high_responders)
        high_res_frontal = [high_res_frontal; data_frontal.(strcat(name,'_power'))(find(data_frontal.subject == high_responders(k)))];
        high_res_parietal = [high_res_parietal; data_parietal.(strcat(name,'_power'))(find(data_parietal.subject == high_responders(k)))];
    end
    
    for y = 1:length(moderate_responders)
        mod_res_frontal = [mod_res_frontal; data_frontal.(strcat(name,'_power'))(find(data_frontal.subject == moderate_responders(y)))];
        mod_res_parietal = [mod_res_parietal; data_parietal.(strcat(name,'_power'))(find(data_parietal.subject == moderate_responders(y)))];
    end
    
    % Average power and place into a table
    info = struct('band',name,...
        'moderate_frontal',mean(mod_res_frontal),...
        'high_frontal',mean(high_res_frontal),...
        'moderate_parietal',mean(mod_res_parietal),...
        'high_parietal',mean(high_res_parietal));
    
    power_info = [power_info,info];
    
end

filename_new = fullfile(output_path,'Component_power.xlsx');
writetable(struct2table(power_info),filename_new);

end

src/functions/mat2excel.m
function [] = mat2excel(input_path,output_path,subj,filename,overwrite)
%% Save EEG into machine learning-ready format (Stage1 -> Stage2)
% PURPOSE: Takes .mat EEG files and saves into a .xlsx file using given
% filename. Each row is a subject and each column is a feature. A feature
% represents a single frequency from the power spectrum density (PSD) from
% 1-100Hz.
%
%   Kyle See 7/1/2021
%   Smart Medical Informatics Learning and Evaluation (SMILE)
%   Biomedical Engineering, University of Florida
%
%   Input:
%       input_path  - Path containing EEG .mat files. Expects a single PSD
%                     per subject.
%       output_path - Path to save .xlsx file.
%
%       Optional Arguments
%       subj        - 1D vector of IDs for each subject. Defaults to 1 
%                     to N IDs (1, 2, 3, 4, 5, ..., N).
%       filename    - Filename for the output .xlsx file. Name needs .xlsx 
%                     extension, otherwise saved as .xls. Default name is
%                     'EEG_features.xlsx'.
%       overwrite   - ['N'|'Y'] 'N' indicates do not overwrite. 'Y'
%                     indicates to overwrite if output file exists. Default
%                     is 'N'.
%
%   Output:
%                   - Generates .xlsx file of PSDs per subject. Each row is
%                     a subject. Each column is a single frequency.
%
%---------------------------------------------
% Last Updated: 7/1/21
% - 7/1 Created
% 
%% Checking inputs and initializing variables
% Retrieving inputs
folder_data = dir(input_path);
dir_data = dir(fullfile(input_path,'*.mat'));

% Check that subj variable and actual files are correct shape
N = length(dir_data); % Number of subjects based on directory
if ~exist('subj','var')
    subj = (1:N)'; % Generates a column vector of unique IDs if not present
elseif length(subj) ~= N
    error('Error. \nThe subj var (n=%d) does not match number of files (n=%d).',length(subj),N)
else
    % Transform to column vector
    if isrow(subj)
        subj = subj';
    end
end

% Checks for filename. Creates one if missing
if ~exist('filename','var')
    filename = 'EEG_features.xlsx';
end

% Checks if need to overwrite
if ~exist('overwrite','var')
    overwrite = 'N';
elseif overwrite ~= 'Y' && overwrite ~= 'N'
    error("Error. \nExpecting overwrite to be 'N' or 'Y'")
end

% Initialize matrix
matrix = [];

%% Convert .mat to .xlsx
% Constructs expected path output
path_output = fullfile(output_path,filename);

% Creates file if it doesn't exist or needs to be overwritten
if isfile(path_output) == 1 && overwrite == 'Y' || isfile(path_output) == 0
    
    % Iterate through input folder path
    for i = 1:length(dir_data)
        
        % Loads .mat file
        file_path = fullfile(input_path,dir_data(i).name);
        data = load(file_path);
        
        % Appends data to matrix
        matrix = [matrix; data.eegData];
    end
    
    % Appends subject ID as first column
    matrix = [subj,matrix];
    
    % Adding headers
    colNames = ["Subject","Hz_" + (1:length(data.eegData))];
    matrix = array2table(matrix,'VariableNames',colNames);
    
    % Function output
    if isfile(path_output) == 1 && overwrite == 'Y'
        writetable(matrix,path_output)
        fprintf("Existing excel file is overwritten.\n")
    elseif isfile(path_output) == 1 && overwrite == 'N'
        fprintf("Current excel file has not been updated.\n")
    elseif isfile(path_output) == 0
        writetable(matrix,path_output)
        fprintf("New excel file is created.\n\n")
    end
end
end

src/functions/eeglab2mat.m
function [] = eeglab2mat(input_path,output_path,overwrite)
%% Save EEG into MATLAB format (Data -> Stage1)
% PURPOSE: Takes EEG data and saves into a .mat file using the same
% file name. The .mat file has size of CHANNEL x TIME x EPOCH matrix.
%
%   Kyle See 7/1/2021
%   Smart Medical Informatics Learning and Evaluation (SMILE)
%   Biomedical Engineering, University of Florida
%
%   Input:
%       input_path  - Path containing EEG .set and .fdt file pairs.
%       output_path - Path to save .mat files. Does not save if file is
%                     already found. Saves variable as "eegData".
%
%       Optional Argument
%       overwrite   - ['N'|'Y'] 'N' indicates do not overwrite. 'Y'
%                     indicates to overwrite if output file exists. Default
%                     is 'N'.
%
%   Output:
%                   - Generates .mat file versions of the EEG .set and .fdt
%                     file pairs. Uses the name of .set and .fdt pair.
%
%---------------------------------------------
% Last Updated: 1/6/22
% - 7/1 Created
% - 1/6 Changed how directory is read. Only reads necessary files instead
%       of parsing through each and every file.

%% Checking inputs and intializing variables
% Checks if need to overwrite
if ~exist('overwrite','var')
    overwrite = 'N';
elseif overwrite ~= 'Y' && overwrite ~= 'N'
    error("Expecting overwrite to be 'N' or 'Y'")
end

% Retrieving inputs
dir_eeg = dir(fullfile(input_path,'*.set'));

% Tracking how many changes
fileCount = 0; % Tracks number of total processed files
newCount = 0; % Tracks new files made
owCount = 0; % Tracks overwritten files

%% Converting .set/.fdt to .mat
% Iterate through input folder path
for i = 1:length(dir_eeg)
    
    % Set up file names
    file = dir_eeg(i).name; % Grabs .set file name of current iteration
    output_name = strcat(extractBefore(file,'.'),'.mat'); % Create .mat file name
    path_output = fullfile(output_path,output_name); % Expected path for saving file
    
    % Perform task if output file does not already exist
    if isfile(path_output) == 0 || isfile(path_output) == 1 && overwrite == 'Y'
        
        % Loading subject with EEGLAB
        EEG = pop_loadset('filename',file,'filepath',input_path,'loadmode','all');
        
        % Extract power spectrum as spectra
        [spectra,~] = spectopo(EEG.data, EEG.pnts, EEG.srate, 'plot','off','verbose','off');
        eegData = mean(spectra,1); % Take mean of spectras across channels
        eegData = eegData(:,1:100); % Keep 1-100Hz frequencies
        fileCount = fileCount + 1;
        
        % Determine overwriting
        if isfile(path_output) == 1 && overwrite == 'Y'
            save(path_output,"eegData")
            owCount = owCount + 1;
        else
            save(path_output,"eegData")
            newCount = newCount + 1;
        end
    elseif isfile(path_output) == 0 || isfile(path_output) == 1 && overwrite == 'N'
        fileCount = fileCount + 1;
    end
end
clc;
fprintf("Detected %d files.\n",fileCount)
fprintf("%d new file(s) created and %d file(s) overwritten.\n\n",newCount,owCount)
end

src/functions/svm_eeg.m
function [] = svm_eeg(input_path,output_path,labels,N,n_fold,feature_sel,pval,overwrite)
%% SVM EEG Channel Classification (Stage2 -> Stage3)
% PURPOSE: Perform n-fold nested cross validation with LIBSVM on EEG
% frequencies across every powerband (All: 1-100Hz, Delta: 1-4Hz, Theta:
% 4-8Hz, Alpha: 8-13Hz, Beta: 13-30Hz, Gamma: 30-100Hz). Creates a final
% model using All data to extract weights for specific frequencies. RNG is
% fixed to default.
%
%   Kyle See 7/1/2021
%   Smart Medical Informatics Learning and Evaluation (SMILE)
%   Biomedical Engineering, University of Florida
%
%   Input:
%       input_path   - Path containing .xlsx file of features.
%       output_path  - Path to save model results.
%       labels       - 1D vector of labels to be used in binary
%                      classification. Must be a vector of 1 and -1.
%
%       Optional Arguments
%       N            - Positive integer for number of iterations. Default
%                      is 10.
%       n_fold       - Positive integer for number of folds. Default is 3.
%       feature_sel  - ['N'|'Y'] 'N' indicates to not use feature
%                      feature selection. 'Y' indicates to use feature
%                      selection t-test. Requires a p-value to use.
%                      Default is 'N'.
%       pval         - Positive number for p-value. Default is 1.
%       overwrite   - ['N'|'Y'] 'N' indicates do not overwrite. 'Y'
%                     indicates to overwrite if output file exists. Default
%                     is 'N'.
%
%   Output:
%       Folder structure containing the entire training process:
%       [] - Indicates folders
%       \\ - Indicates a file
%       -------------------------------------
%       [output_path]
%           [outerfold_x]*
%               [innerfold_y]**
%                   \model.mat\
%           \acc.mat\
%           \final_model.mat\
%       -------------------------------------
%       model.mat: contains various model results. Descriptions for each
%                variable is in the train_svm.m file.
%       acc.mat: gives the accuracy and confidence intervals for training
%                and testing.
%       final_model.mat: the final trained SVM model.
%
%       *outerfold has N many folders numbered with x
%       **innerfold has n_fold many folders numbered with y
%
%   Dependencies:
%       LIBSVM-3.24
%
%---------------------------------------------
% Last Updated: 7/1/21
% - 7/1 Created
% - 7/21 Removed band preference. Now performs SVM using every band.

%% Checking inputs and initializing variables
% Check for iterations
if ~exist('N','var')
    N = 10;
elseif  N <= 0 && floor(N) ~= N
    error("N is not a positive integer. N is %d",N)
end

% Check for number of folds
if ~exist('n_fold','var')
    n_fold = 3;
elseif n_fold <= 0 && floor(n_fold) ~= n_fold
    error("n_fold is not a positive integer")
end

% Check for feature selection
if ~exist('feature_sel','var')
    feature_sel = 'N';
elseif feature_sel ~= 'Y' && feature_sel ~= 'N'
    error("Expecting feature_sel to be 'N' or 'Y'")
end

% Check p-value
if ~exist('pval','var') || strcmp(feature_sel,'N')
    pval = 1;
elseif pval < 0
    error("pval is not a positive number")
end

% Save the results
if ~exist('overwrite','var')
    overwrite = 'N';
elseif overwrite ~= 'N' && overwrite ~= 'Y'
    error("Expected save_result to be 'N' or 'Y'")
end

% Initialize variables
acc_train = zeros(N, n_fold); % Training accuracy
acc_test = zeros(N, n_fold); % Testing accuracy
C = zeros(N, n_fold); % C variable in LIBSVM
unitag = datestr(now,30); % Unitag
bands = ["All", "Delta", "Theta", "Alpha", "Beta" ,"Gamma"];

%% Selecting powerband
% Extract x
folder_data = dir(input_path);
file = folder_data(3).name;
data = readtable(fullfile(input_path,file));
x = data{:,2:end};

% Changes to column vector
if isrow(labels)
    labels = labels';
end

% Repeat for each band
for i = 1:6
    % Band selection
    if strcmp(bands(i),'Delta')
        X = x(:,1:4);
    elseif strcmp(bands(i),'Theta')
        X = x(:,4:8);
    elseif strcmp(bands(i),'Alpha')
        X = x(:,8:13);
    elseif strcmp(bands(i),'Beta')
        X = x(:,13:30);
    elseif strcmp(bands(i),'Gamma')
        X = x(:,30:end);
    else
        X = x;
    end
    
    % Expected Band output path
    output_path_band = fullfile(output_path,bands(i));
    
    % Create folder if it does not exist
    if ~exist(output_path_band,'dir')
        mkdir(output_path_band)
    end
    
%% Machine learning model
    rng default
    % Repeat for N iterations
    for repeat = 1:N
        cv1 = cvpartition(labels,'Kfold',n_fold,'Stratify',true); % Redefines CV per iteration
        
        % Repeat for n_folds
        for fold = 1:n_fold
            
            % Train SVM
            perf = train_svm(X,labels,pval,cv1,fold,repeat,output_path_band,unitag,feature_sel,overwrite);
            
            % Collect results
            acc_train(repeat,fold) = perf.acc_train;
            acc_test(repeat,fold) = perf.acc_test;
            C(repeat,fold) = perf.C;
            
            if fold == 1
                targs = perf.labels(perf.testIdx);
                preds = perf.ypred_test;
            else
                targs = [targs; perf.labels(perf.testIdx)];
                preds = [preds; perf.ypred_test];
            end
        end
        
        % Metrics
        TP(repeat) = sum(targs == 1 & preds == 1); % True Positive
        FP(repeat) = sum(targs == -1 & preds == 1); % False Positive
        TN(repeat) = sum(targs == -1 & preds == -1); % True Negative
        FN(repeat) = sum(targs == 1 & preds == -1); % False Negative
    end
    
    % Save Output
    output.acc_train = acc_train;
    output.avg_acc_train = mean(mean(acc_train));
    output.acc_test = acc_test;
    output.avg_acc_test = mean(mean(acc_test));
    
    % Average Accuracy + 95% Conf Interval
    ACC_train = reshape(acc_train, [(N*n_fold) 1]);
    ACC_test = reshape(acc_test, [(N*n_fold) 1]);
    cip_train = mean(ACC_train) + (1.96*(std(ACC_train)/sqrt(length(ACC_train))));
    cin_train = mean(ACC_train) - (1.96*(std(ACC_train)/sqrt(length(ACC_train))));
    cip_test = mean(ACC_test) + (1.96*(std(ACC_test)/sqrt(length(ACC_test))));
    cin_test = mean(ACC_test) - (1.96*(std(ACC_test)/sqrt(length(ACC_test))));
    
    % Save + Print Results
    Accuracy_train = mean(ACC_train);
    Accuracy_test = mean(ACC_test);
    Acc = mean(acc_test,2)';
    C_optimal = mode(mode(C));
    
    if overwrite == 'Y'
        save(fullfile(output_path_band,['acc.mat']),'cin_train','cip_train','Accuracy_train','cin_test','cip_test','Accuracy_test','acc_train','acc_test')
    end
    
    fprintf('Repeated Nested Cross-Validation Results\n')
    fprintf('Train Accuracy CI: [%.2f%% - %.2f%%]\n', cin_train*100, cip_train*100);
    fprintf('Train Accuracy: %.2f\n', Accuracy_train*100);
    fprintf('Test Accuracy CI: [%.2f%% - %.2f%%]\n', cin_test*100, cip_test*100);
    fprintf('Test Accuracy: %.2f\n\n', Accuracy_test*100);
    
%% Final Model
    % Feature Selection
    if feature_sel == 1
        [~,p] = ttest2(X(labels == 1,:), X(labels == 0,:));
        features = (p < pval)';
        X = X(:,features==1);
    end
    
    % Normalize data
    X = (X - mean(X,1)) ./ std(X,0,1);
    model = svmtrain(labels,X,['-t 0 -h 1 -q -c ', num2str(C_optimal)]);
 
%% Saving Results
    if overwrite == 'Y'
        if feature_sel == 1
            save(fullfile(output_path_band,'final_model.mat'),'model','feature_sel')
        else
            save(fullfile(output_path_band,'final_model.mat'),'model')
        end
    end
end
%% Find component weights from CV
band_path = dir(output_path);
band_path = band_path(~ismember({band_path.name},{'.','..'})); % Remove . and .. from directory

% Iterate through each Powerband Folder
for ii = 1:length(band_path)
    
    % Create directory for iterations
    Outer_name = fullfile(output_path,band_path(ii).name); % Name for powerband folder
    Outer_storage = dir(Outer_name); % Directory for powerband folder
    Outer_storage = Outer_storage(~ismember({Outer_storage.name},{'.','..'})); % Remove . and .. from directory
    
    % Reset weight vector before new band
    weight = [];
    
    % Iterate through each iteration
    for jj = 4:length(Outer_storage) % Assumes acc,cross_val,and final are in the folder
        
        % Create directory for cv folds
        Inner_name = fullfile(Outer_name,Outer_storage(jj).name);
        Inner_storage = dir(Inner_name);
        Inner_storage = Inner_storage(~ismember({Inner_storage.name},{'.','..'})); % Remove . and .. from directory
        
        % Iterate through each cv fold
        for kk = 1:length(Inner_storage)
            
            % Load model for a single fold
            Model_name = fullfile(Inner_name,Inner_storage(kk).name,'model.mat');
            load(Model_name)
            
            % Create or append existing weight
            if strcmp(band_path(ii).name,'All')
                if jj == 4 && kk == 1
                    % Initialize weight
                    weight = perf.weights;
                else
                    % Take average of weights
                    weight = (weight+perf.weights)/2;
                end
            elseif strcmp(band_path(ii).name,'Alpha')
                if jj == 4 && kk == 1
                    weight = perf.weights;
                else
                    weight = (weight+perf.weights)/2;
                end
            elseif strcmp(band_path(ii).name,'Beta')
                if jj == 4 && kk == 1
                    weight = perf.weights;
                else
                    weight = (weight+perf.weights)/2;
                end
            elseif strcmp(band_path(ii).name,'Delta')
                if jj == 4 && kk == 1
                    weight = perf.weights;
                else
                    weight = (weight+perf.weights)/2;
                end
            elseif strcmp(band_path(ii).name,'Gamma')
                if jj == 4 && kk == 1
                    weight = perf.weights;
                else
                    weight = (weight+perf.weights)/2;
                end
            elseif strcmp(band_path(ii).name,'Theta')
                if jj == 4 && kk == 1
                    weight = perf.weights;
                else
                    weight = (weight+perf.weights)/2;
                end
            end
        end
    end
    save(fullfile(Outer_name,'cross_val.mat'),'weight')
end

end